### PR TITLE
try to port ASCIIUtility.NarrowUtf16ToAscii to x-plat intrinsics

### PR DESCRIPTION
**Pull request by adamsitnik**
For both x64 and ARM64 I am observing 10-15% regression.

### x64

<details>

```ini
BenchmarkDotNet=v0.13.1.1828-nightly, OS=Windows 11 (10.0.22000.795/21H2)
AMD Ryzen Threadripper PRO 3945WX 12-Cores, 1 CPU, 24 logical and 12 physical cores
.NET SDK=7.0.100-preview.6.22352.1
  [Host]     : .NET 7.0.0 (7.0.22.32404), X64 RyuJIT AVX2
  
Method=GetBytes
```

|              Type |  Job | size | encName |              Input |          Mean | Ratio |
|------------------ |----- |----- |-------- |------------------- |--------------:|------:|
|     Perf_Encoding |   PR |   16 |   ascii |                  ? |      18.22 ns |  1.00 |
|     Perf_Encoding | main |   16 |   ascii |                  ? |      18.20 ns |  1.00 |
|                   |      |      |         |                    |               |       |
|     Perf_Encoding |   PR |   16 |   utf-8 |                  ? |      17.81 ns |  1.09 |
|     Perf_Encoding | main |   16 |   utf-8 |                  ? |      16.63 ns |  1.00 |
|                   |      |      |         |                    |               |       |
|     Perf_Encoding |   PR |  512 |   ascii |                  ? |      50.45 ns |  1.12 |
|     Perf_Encoding | main |  512 |   ascii |                  ? |      44.88 ns |  1.00 |
|                   |      |      |         |                    |               |       |
|     Perf_Encoding |   PR |  512 |   utf-8 |                  ? |      74.30 ns |  1.16 |
|     Perf_Encoding | main |  512 |   utf-8 |                  ? |      64.30 ns |  1.00 |
|                   |      |      |         |                    |               |       |
| Perf_Utf8Encoding |   PR |    ? |       ? |    EnglishAllAscii |  17,846.09 ns |  1.08 |
| Perf_Utf8Encoding | main |    ? |       ? |    EnglishAllAscii |  16,454.56 ns |  1.00 |
|                   |      |      |         |                    |               |       |
| Perf_Utf8Encoding |   PR |    ? |       ? | EnglishMostlyAscii |  79,019.77 ns |  0.97 |
| Perf_Utf8Encoding | main |    ? |       ? | EnglishMostlyAscii |  81,128.97 ns |  1.00 |
|                   |      |      |         |                    |               |       |
| Perf_Utf8Encoding |   PR |    ? |       ? |            Chinese |  82,925.92 ns |  0.99 |
| Perf_Utf8Encoding | main |    ? |       ? |            Chinese |  83,539.51 ns |  1.00 |
|                   |      |      |         |                    |               |       |
| Perf_Utf8Encoding |   PR |    ? |       ? |           Cyrillic |  97,716.52 ns |  0.98 |
| Perf_Utf8Encoding | main |    ? |       ? |           Cyrillic |  99,311.72 ns |  1.00 |
|                   |      |      |         |                    |               |       |
| Perf_Utf8Encoding |   PR |    ? |       ? |              Greek | 151,446.91 ns |  0.96 |
| Perf_Utf8Encoding | main |    ? |       ? |              Greek | 157,171.66 ns |  1.00 |

</details>

Here I was able to use VTune (I also tried uProf but it leaves a lot to desire):

vectors_vtune

From what I can see the regression comes from additional 3 `vpand` instructions:

image

It seems that the first one comes from the new `VectorContainsNonAsciiChar` implementation (this is expected). But the other two from `Vector128.Narrow`? @tannergooding is that expected?

### ARM64

<details>

```ini
BenchmarkDotNet=v0.13.1.1828-nightly, OS=ubuntu 20.04
Unknown processor
.NET SDK=7.0.100-rc.1.22379.1
  [Host]     : .NET 7.0.0 (7.0.22.37802), Arm64 RyuJIT AdvSIMD

Method=GetBytes
```

|              Type |  Job | size | encName |              Input |          Mean | Ratio |
|------------------ |----- |----- |-------- |------------------- |--------------:|------:|
|     Perf_Encoding |   PR |   16 |   ascii |                  ? |      67.68 ns |  0.98 |
|     Perf_Encoding | main |   16 |   ascii |                  ? |      68.75 ns |  1.00 |
|                   |      |      |         |                    |               |       |
|     Perf_Encoding |   PR |   16 |   utf-8 |                  ? |      75.34 ns |  1.00 |
|     Perf_Encoding | main |   16 |   utf-8 |                  ? |      75.26 ns |  1.00 |
|                   |      |      |         |                    |               |       |
|     Perf_Encoding |   PR |  512 |   ascii |                  ? |     202.77 ns |  1.12 |
|     Perf_Encoding | main |  512 |   ascii |                  ? |     180.61 ns |  1.00 |
|                   |      |      |         |                    |               |       |
|     Perf_Encoding |   PR |  512 |   utf-8 |                  ? |     269.92 ns |  1.13 |
|     Perf_Encoding | main |  512 |   utf-8 |                  ? |     239.16 ns |  1.00 |
|                   |      |      |         |                    |               |       |
| Perf_Utf8Encoding |   PR |    ? |       ? |    EnglishAllAscii |  79,699.39 ns |  1.15 |
| Perf_Utf8Encoding | main |    ? |       ? |    EnglishAllAscii |  69,283.09 ns |  1.00 |
|                   |      |      |         |                    |               |       |
| Perf_Utf8Encoding |   PR |    ? |       ? | EnglishMostlyAscii | 285,553.35 ns |  1.00 |
| Perf_Utf8Encoding | main |    ? |       ? | EnglishMostlyAscii | 285,896.31 ns |  1.00 |
|                   |      |      |         |                    |               |       |
| Perf_Utf8Encoding |   PR |    ? |       ? |            Chinese | 308,349.96 ns |  1.00 |
| Perf_Utf8Encoding | main |    ? |       ? |            Chinese | 307,750.04 ns |  1.00 |
|                   |      |      |         |                    |               |       |
| Perf_Utf8Encoding |   PR |    ? |       ? |           Cyrillic | 217,887.57 ns |  1.00 |
| Perf_Utf8Encoding | main |    ? |       ? |           Cyrillic | 217,912.29 ns |  1.00 |
|                   |      |      |         |                    |               |       |
| Perf_Utf8Encoding |   PR |    ? |       ? |              Greek | 312,137.16 ns |  1.00 |
| Perf_Utf8Encoding | main |    ? |       ? |              Greek | 311,067.67 ns |  1.00 |

</details>

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs [hunk 1]
```diff
@@ -1471,7 +1410,7 @@ private static unsafe nuint NarrowUtf16ToAscii_Intrinsified(char* pUtf16Buffer,
             // Turn the 8 ASCII chars we just read into 8 ASCII bytes, then copy it to the destination.
 
             ref byte asciiBuffer = ref *pAsciiBuffer;
-            Vector128<byte> asciiVector = ExtractAsciiVector(utf16VectorFirst, utf16VectorFirst);
+            Vector128<byte> asciiVector = Vector128.Narrow(utf16VectorFirst, utf16VectorFirst);
```
Comment (EgorBo) [inline src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs:1413]: @adamsitnik Narrow doesn't know that `utf16VectorFirst` is already ASCII at this point (from my understanding) so it applies a mask via AND to cut anything above 0xFF. Consider this:
image

So for this case for better perf you probably want to keep `ExtractAsciiVector`

File: src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs [hunk 2]
```diff
@@ -1380,64 +1373,10 @@ public static unsafe nuint NarrowUtf16ToAscii(char* pUtf16Buffer, byte* pAsciiBu
...
-                return AdvSimd.Arm64.UnzipEven(vectorFirst.AsByte(), vectorSecond.AsByte());
-            }
-            else
-            {
-                throw new PlatformNotSupportedException();
-            }
+            const ushort asciiMask = ushort.MaxValue - 127; // 0x7F80
+            Vector128<ushort> zeroIsAscii = utf16Vector & Vector128.Create(asciiMask);
+            // If a non-ASCII bit is set in any WORD of the vector, we have seen non-ASCII data.
+            return !Vector128.EqualsAll(zeroIsAscii, Vector128<ushort>.Zero);
```
Comment (EgorBo) [inline src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs:1379]: You said that it's expected to have a redundant AND here - why? 🙂

**Comment by adamsitnik** [comment]
@tannergooding Since I was not able to get the same perf with `Vector128`, I've added `Vector128` as a fallback when SSE2 and AdvSIMD are not supported which in theory some configs may benefit from.

**Review by tannergooding — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs (hunk 1 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs]: Can we log an issue here for `Narrow`. There is possibly some codegen improvements we can make in .NET 8

**Comment by kunalspathak** [comment]
Improvements - https://github.com/dotnet/perf-autofiling-issues/issues/7322

**Pull request merged**